### PR TITLE
Adds full support for new Bioicons library

### PR DESCRIPTION
**Pull request by mattwarkentin**
Hi @mitchelloharawild,

First off, thanks for creating this great package. [A few days ago a new icon library was released](https://twitter.com/simonduerr/status/1385852166804545536) and it has a really rich set of biological/science focused icons. I was thinking about spinning it into a package, but I think it makes much more sense for it to join the growing `icons` eco-system.

Bioicons Website: https://bioicons.com/
Bioicons GitHub: https://github.com/duerrsimon/bioicons

Hopefully you are interested in this PR and I am happy to refine the PR to make any desired changes.

### CODE REVIEW

**Comment by mattwarkentin** [comment]
Also, the rendered `README.md` got a little funky because I rendered it without having all of the icon libraries installed. I tried re-running it after installing the missing icon sets, but `octicons` failed to install, with the following error:
```
trying URL 'https://github.com/primer/octicons/archive/master.zip'
Error in download.file(url, dl_file) : 
  cannot open URL 'https://github.com/primer/octicons/archive/master.zip'
In addition: Warning message:
In download.file(url, dl_file) :
  cannot open URL 'https://codeload.github.com/primer/octicons/zip/master': HTTP status was '404 Not Found'
```

You may need to render the README again on your end.

**Comment by mattwarkentin** [comment]
Actually, I'm going to make a big update to this PR to adhere closer to the other icon sets. Stay tuned.

File: DESCRIPTION [hunk 1]
```diff
@@ -22,7 +23,7 @@ Imports:
     cli,
     crayon,
     utils,
-    rstudioapi
+    rstudioapi,
```
Comment (mitchelloharawild) [inline DESCRIPTION:26]: I've removed this hanging ,

File: R/bioicons.R [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    "bioicons", url, c("static", "icons"),
+    meta = list(
+      name = "Bioicons",
+      version = meta$version,
+      licence = c("BSD", "CC-0", "CC-BY-SA", "MIT")
+    )
+  )
+  saveRDS(
+    icon_df,
+    glue::glue("{rappdirs::user_data_dir('rpkg_icon/bioicons')}/icons.rds")
```
Comment (mitchelloharawild) [inline R/bioicons.R:28]: I've changed this to use the `icon_path()` function. It allows users to specify where they want the icons stored, and also resolves issues with `/` not always being the appropriate file separator.

**Review by mitchelloharawild — APPROVED** [review]
Looks great, thanks!
The bioicons repo is interesting in how each icon has its own license and author. The {icons} package would need extending to properly support this. I think for now it is simplest to set the license to something like 'various' and provide a link to <https://bioicons.com/> which provides attribution and licensing requirements. What do you think of this approach?

I've fixed a few small issues in the PR, and accidentally merged in the process.
I see that you've made some nice customisations in the `bioicons()` function allowing direct access to icons with code like: `bioicons("waterfall")` - very nice.

I wonder what you think about the alternative `$` interface, and if it could be improved. The equivalent code to get the waterfall icon would be: `bioicons$mit$Scientific_graphs$`dbs-sticky`$waterfall`. Dropping the attribution would make this `bioicons$Scientific_graphs$waterfall`. Can the attributions be later obtained from `icons.Rds` based only on the category and name? I imagine that the name is a unique identifier for the icon.

File: DESCRIPTION (hunk 1 above)
Comment (mattwarkentin) [reply DESCRIPTION:26]: 🤦‍♂️

**Comment by mattwarkentin** [comment]
> I think for now it is simplest to set the license to something like 'various' and provide a link to https://bioicons.com/ which provides attribution and licensing requirements. What do you think of this approach?

I think that makes sense. I did list all current licences used by Bioicons in the meta data.

https://github.com/mitchelloharawild/icons/blob/a9c28432c8fcb55135d63aa6ff916a27842e18cd/R/bioicons.R#L20-L24

> I wonder what you think about the alternative `$` interface, and if it could be improved.

Ah yes, I didn't pay too much attention to this interface to the icons. Having to navigate through the attribution and author is definitely not ideal and will make it hard to find a specific icon that way. You are probably right, we should probably remove the attribution and authors to make it more useful.

> I imagine that the name is a unique identifier for the icon.

Unfortunately the icon name is not unique. This is why I did a little extra work to make sure that, by default, an icon will "just work" by providing a name, but that users can still provide the category if they want a specific "version" of the icon.  Thankfully there are no duplicate names _within_ a category, but there is nothing to guarantee this stays true in the future.

There are 30 cases of duplicate icon names. For example, when querying "gpu", there are two matches by name <https://bioicons.com/?query=gpu>. This is sort of annoying and perhaps the `Bioicons` maintainer might consider making all icon names unique if we asked nicely.

**Comment by mitchelloharawild** [comment]
> Unfortunately the icon name is not unique. This is why I did a little extra work to make sure that, by default, an icon will "just work" by providing a name, but that users can still provide the category if they want a specific "version" of the icon. Thankfully there are no duplicate names _within_ a category, but there is nothing to guarantee this stays true in the future.
> 
> There are 30 cases of duplicate icon names. For example, when querying "gpu", there are two matches by name https://bioicons.com/?query=gpu. This is sort of annoying and perhaps the `Bioicons` maintainer might consider making all icon names unique if we asked nicely.

Thanks for this info, I'll try to restructure the directory to allow `bioicons$category$name`, given that this combination is a unique identifier. This should also remove the need to replace spaces with underscores for the author names as far as I can tell.

I also noticed that your `bioicons()` function assumes that `name` has a valid match, due to how it indexes `chosen_icon`.
I've fixed this in https://github.com/mitchelloharawild/icons/commit/243b2d7f9cda790f2ba2f8b897730506b279c698


``` r
library(icons)
#> ── Installed icons ─────────────────────────────────────────────── icon 0.2.0 ──
#> ✓ ionicons        5.2.3       ✓ google_material 4.0.0  
#> ✓ academicons     1.9.1       ✓ feather_icons   v4.28.0
#> ✓ simple_icons    4.4.0       ✓ octicons        11.2.0 
#> ✓ bioicons        1.0.0       ✓ fontawesome     5.15.2
bioicons("?")
#> Error: The `?` icon could not be found in this icon set.
```

<sup>Created on 2021-04-27 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by mattwarkentin** [comment]
I am unclear as to what that error is meant to be showing

**Comment by mitchelloharawild** [comment]
I realised just now that the reprex was wrong. :see_no_evil: 

**Before:**
``` r
library(icons)
#> ── Installed icons ─────────────────────────────────────────────── icon 0.2.0 ──
#> ✓ ionicons        5.2.3       ✓ google_material 4.0.0  
#> ✓ academicons     1.9.1       ✓ feather_icons   v4.28.0
#> ✓ simple_icons    4.4.0       ✓ octicons        11.2.0 
#> ✓ bioicons        1.0.0       ✓ fontawesome     5.15.2
bioicons("?")
#> Error in chosen_icon$license[[1]]: subscript out of bounds
```

<sup>Created on 2021-04-27 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**After https://github.com/mitchelloharawild/icons/commit/243b2d7f9cda790f2ba2f8b897730506b279c698:**
``` r
library(icons)
#> ── Installed icons ─────────────────────────────────────────────── icon 0.2.0 ──
#> ✓ ionicons        5.2.3       ✓ google_material 4.0.0  
#> ✓ academicons     1.9.1       ✓ feather_icons   v4.28.0
#> ✓ simple_icons    4.4.0       ✓ octicons        11.2.0 
#> ✓ bioicons        1.0.0       ✓ fontawesome     5.15.2
bioicons("?")
#> Error: The `?` icon could not be found in this icon set.
```

<sup>Created on 2021-04-27 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by duerrsimon** [comment]
Chiming in here. Cool to see the library integrated this quickly into another package :) 

I'll fix the duplicate icon names as soon as I have time. 

Attributing the icons to bioicons.com though is clearly not in line with the CC BY/SA licenses and it would be better if you could provide an automatically generated attribution like the bioicons.com website now also does: 
e.g `hepatits-virus icon by Servier https://smart.servier.com/ is licensed under CC-BY 3.0 Unported https://creativecommons.org/licenses/by/3.0/`

**Comment by mitchelloharawild** [comment]
Thanks @duerrsimon for your chiming :+1: :wind_chime:

Regarding attribution and this package, it is currently expected that the user would correctly attribute the used icon. More tools to encourage users to correctly attribute works is wanted, and the current plan is here: https://github.com/mitchelloharawild/icons/issues/15

Users usually attribute icon use at the end of the document, as the usual icons are small and part of the interface/design. I expect the usage of bioicons would be different though, as the graphics are more detailed and would likely be more appropriately used in/as larger graphics. I can justify more suggestive licensing messages when bioicons are used for this reason.

Another limiting assumption made by this package currently is that all icons within an icon set have the same license. This will need to be generalised to properly support #15 and display the appropriate attribution requirements.

**Comment by duerrsimon** [comment]
Just a little heads up. I fixed the duplicate icons and added some more icons :) 
Would it help you that whenever new icons are added a new release tag is created such that it is trackable whether the package is in sync with the library?

**Pull request merged**